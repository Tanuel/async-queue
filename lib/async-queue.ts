export class AsyncQueue {
  /**
   * Number of resolved jobs
   */
  private resolved = 0;
  /**
   * Number of rejected jobs
   */
  private rejected = 0;
  /**
   * Number of pending jobs, tracked for limiting
   */
  private pending = 0;
  /**
   * Array pointer
   */
  private current = 0;
  /**
   * Max amount of jobs to be running "simultaneously" (as far as javascript/threading allows this term)
   * Can be changed anytime.
   */
  public limit: number;
  /**
   * Array of Jobs
   */
  private readonly jobs: CallableFunction[] = [];
  /**
   * Container for all callbacks for the different ListenerTypes
   * @see ListenerTypes
   */
  private callbacks: CallbackContainer = {
    next: [],
    reject: [],
    done: [],
    resolve: [],
    pending: [],
    push: [],
    finally: [],
  };
  constructor(options?: AsyncQueueOptions) {
    this.limit = options?.limit ?? 0;
    this.jobs = options?.jobs ?? [];
    this.next();
  }

  /**
   * When a job is rejected
   * @see OnReject
   */
  private reject(err: any) {
    this.callbacks.reject.forEach((cb) =>
      cb.call(this, err, this.current, this.jobs.length)
    );
    this.rejected++;
    this.finally(err);
  }

  /**
   * When a job is resolved
   * @see OnResolve
   */
  private resolve(result: any) {
    this.callbacks.resolve.forEach((cb) =>
      cb.call(this, result, this.current, this.jobs.length)
    );
    this.resolved++;
    this.finally(result);
  }

  /**
   * When a pending job is done, decrement counter and trigger next
   * @see OnFinally
   */
  private finally(result: any) {
    this.callbacks.finally.forEach((cb) =>
      cb.call(this, result, this.current, this.jobs.length)
    );

    this.pending--;
    if (this.current >= this.jobs.length && this.pending === 0) {
      this.callbacks.done.forEach((cb) => cb.call(this, this.jobs.length));
    } else if (this.current >= this.jobs.length) {
      this.callbacks.pending.forEach((cb) => cb.call(this, this.pending));
    } else {
      this.next();
    }
  }

  /**
   * Put a job in the queue and return the result when done
   */
  public async push<T>(fn: CallableFunction): Promise<T> {
    this.callbacks.push.forEach((cb) =>
      cb.call(this, fn, this.current, this.jobs.length)
    );
    return new Promise<T>((resolve, reject) => {
      this.jobs.push(async () => {
        try {
          const result = await fn();
          resolve(result);
          return result;
        } catch (e) {
          reject(e);
          throw e;
        }
      });
      this.next();
    });
  }

  /**
   * Put multiple jobs in the queue and return Promise.all
   */
  public async pushMulti(...cbs: CallableFunction[]) {
    const promises = cbs.map((cb) => {
      return new Promise((resolve) => {
        this.jobs.push(async () => {
          resolve(await cb());
        });
        this.next();
      });
    });
    return Promise.all(promises);
  }

  /**
   * Trigger the next job (if not over limit)
   */
  private next(): boolean {
    // check if limit is enabled or reached
    if (this.limit > 0 && this.pending >= this.limit) {
      return false;
    }
    // get next job
    const job = this.jobs[this.current];
    if (!job) {
      return false;
    }
    // increment counters for tracking
    this.pending++;
    this.current++;

    // trigger next job asynchronously
    runDelayed(job).then(
      (result) => this.resolve(result),
      (err) => this.reject(err)
    );
    // trigger onNext callbacks
    this.callbacks.next.forEach((cb) =>
      cb.call(this, this.current, this.jobs.length)
    );
    // trigger next job
    this.next();
    return true;
  }

  /**
   * Add an event listener for several types of events
   * @see ListenerTypes
   * @see CallbackContainer
   */
  // FIXME: Add typing for Callback types depending on ListenerTypes
  // public on<T extends ListenerTypes>(listener: T, callback: Callback<T>): AsyncQueue;
  public on(
    listener: ListenerTypes,
    callback: (...params: any[]) => any
  ): AsyncQueue {
    this.callbacks[listener].push(callback);
    return this;
  }
}

/**
 * Helper function to wrap a callback in a promise
 */
function runDelayed(cb: CallableFunction) {
  return new Promise((resolve, reject) => {
    setTimeout(async () => {
      try {
        const result = await cb();
        resolve(result);
      } catch (e) {
        reject(e);
      }
    });
  });
}

/**
 * Next Job has been started
 */
export interface OnNext {
  (this: AsyncQueue, current: number, length: number): any;
}

/**
 * One or multiple jobs have been added to the queue
 */
export interface OnPush {
  (this: AsyncQueue, callback: any, current: number, length: number): any;
}

/**
 * A Job has been resolved
 */
export interface OnResolve {
  (this: AsyncQueue, result: any, current: number, length: number): any;
}

/**
 * A Job has been rejected
 */
export interface OnReject {
  (this: AsyncQueue, error: any, current: number, length: number): any;
}

/**
 * A Job has either been resolved or rejected
 */
export interface OnFinally {
  (this: AsyncQueue, result: any, current: number, length: number): any;
}

/**
 * All Jobs have been finished
 */
export interface OnDone {
  (this: AsyncQueue, length: number): any;
}

/**
 * No Jobs are left in the queue, but jobs are still pending completion
 */
export interface OnPending {
  (this: AsyncQueue, pending: number): any;
}

/**
 * The different types of event listeners
 */
type ListenerTypes =
  | "reject"
  | "done"
  | "resolve"
  | "pending"
  | "push"
  | "next"
  | "finally";

interface CallbackContainer {
  next: OnNext[];
  reject: OnReject[];
  done: OnDone[];
  resolve: OnResolve[];
  pending: OnPending[];
  push: OnPush[];
  finally: OnFinally[];
}

// FIXME: Add typing for Callback types depending on ListenerTypes
// type Callback<T extends ListenerTypes> = T extends Array<infer U> ? U : CallbackContainer[T];

/**
 * Options to pass to AsyncQueue
 */
export interface AsyncQueueOptions {
  /**
   * Max number of jobs running at once (whatever that means in single-threaded javascript)
   */
  limit?: number;
  /**
   * Array of initial functions to start of the queue
   */
  jobs?: CallableFunction[];
}
