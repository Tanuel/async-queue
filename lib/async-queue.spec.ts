import { AsyncQueue } from "./async-queue";

describe("AsyncQueue", () => {
  test("resolve", async () => {
    const queue1 = new AsyncQueue();
    const fn = () =>
      new Promise((resolve) => setTimeout(resolve.bind(null, "resolved"), 100));
    const result = await queue1.push(fn);
    expect(result).toBe("resolved");
  });

  test("reject", async () => {
    const queue2 = new AsyncQueue();
    const fn = () =>
      new Promise((resolve, reject) =>
        setTimeout(reject.bind(null, "rejected"), 100)
      );
    try {
      await queue2.push(fn);
    } catch (e) {
      expect(e).toBe("rejected");
    }
  });

  test("hooks", async () => {
    const mockPush = jest.fn();
    const mockResolve = jest.fn();
    const mockReject = jest.fn();
    const mockFinally = jest.fn();
    const mockNext = jest.fn();
    const mockPending = jest.fn();
    const mockDone = jest.fn();
    const queue3 = new AsyncQueue();
    // wrap in a promise to resolve later when everything is done
    const awaiter = new Promise((resolve) => {
      queue3
        .on("push", (response) => mockPush(response))
        .on("resolve", (response) => mockResolve(response))
        .on("reject", (response) => mockReject(response))
        .on("finally", (response) => mockFinally(response))
        .on("next", (response) => mockNext(response))
        .on("pending", (response) => {
          mockPending(response);
        })
        .on("done", (response) => mockDone(response))
        // this will be used later to await all pending jobs
        .on("done", () => {
          resolve();
        });
    });
    const fnResolve = () =>
      new Promise((resolve) => setTimeout(resolve.bind(null, "resolved"), 10));
    const fnReject = () =>
      new Promise((resolve, reject) =>
        setTimeout(reject.bind(null, "rejected"), 10)
      );
    // ignore result here, since we will handle it in the resolve hook
    queue3.push(fnResolve);
    // ignore errors, but node requires catching rejections
    queue3.push(fnReject).catch(() => null);

    // wait for queue to finish all calls in the node event loop
    await awaiter;

    expect(mockPush).toHaveBeenCalledWith(fnResolve);
    expect(mockPush).toHaveBeenCalledWith(fnReject);
    expect(mockResolve).toHaveBeenCalledWith("resolved");
    expect(mockReject).toHaveBeenCalledWith("rejected");
    expect(mockFinally).toHaveBeenCalledWith("resolved");
    expect(mockFinally).toHaveBeenCalledWith("rejected");
    expect(mockPending).toHaveBeenCalledWith(1);
    expect(mockPending).toHaveBeenCalledTimes(1);
    expect(mockDone).toHaveBeenCalledWith(2);
    expect(mockDone).toHaveBeenCalledTimes(1);
  });

  test("push multi", async () => {
    const jobs = [];
    for (let i = 0; i < 10; i++) {
      jobs.push(() => new Promise((resolve) => resolve(i)));
    }

    const queue4 = new AsyncQueue();
    const result = await queue4.pushMulti(...jobs);
    expect(result).toMatchSnapshot();
  });

  // Fake a mutex to check if only one job is running at a time
  test("limit", async () => {
    const mutexError = jest.fn();
    const mutexTriggered = jest.fn();
    const mutexReleased = jest.fn();
    const jobs = [];
    let mutex = false;
    const requestMutex = (num: any) => {
      if (mutex) {
        mutexError(num);
      }
      mutexTriggered();
      mutex = true;
    };
    const releaseMutex = (num: any) => {
      mutex = false;
      mutexReleased();
    };
    for (let i = 0; i < 10; i++) {
      jobs.push(
        () =>
          new Promise((resolve) => {
            requestMutex(i);
            setTimeout(() => {
              releaseMutex(i);
              resolve(i);
            }, 10);
          })
      );
    }

    const queue4 = new AsyncQueue({
      limit: 1,
    });
    await queue4.pushMulti(...jobs);
    expect(mutexError).not.toHaveBeenCalled();
    expect(mutexTriggered).toHaveBeenCalledTimes(10);
    expect(mutexReleased).toHaveBeenCalledTimes(10);
  });
});
