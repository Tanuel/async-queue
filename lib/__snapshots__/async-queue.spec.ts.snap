// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncQueue push multi 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
]
`;
